pennylane/spin/spin_hamiltonian.py

```diff
@@ -298,3 +309,347 @@ def fermi_hubbard(
+            a square matrix of size ``(num_spins, num_spins)``, where ``num_spins`` is the total
+            number of spins. Default value is 1.0.
+        phi (float or List[math.array(float)]): Phase Factor, it can be a number, or
+            a square matrix of size ``(num_spins, num_spins)``, where ``num_spins`` is the total
+            number of spins. Default value is 1.0.
+        boundary_condition (bool or list[bool]): Defines boundary conditions for different lattice
+            axes, default is ``False`` indicating open boundary condition.
+        mapping (str): Specifies the fermion-to-qubit mapping. Input values can be
+            ``'jordan_wigner'``, ``'parity'`` or ``'bravyi_kitaev'``.
+
+    Returns:
+       pennylane.operator: Hamiltonian for the Haldane model.
+
+    **Example**
+
+    >>> n_cells = [2]
+    >>> h = [0.5]
+    >>> u = 1.0
+    >>> v = 0.2
+    >>> spin_ham = qml.spin.haldane("chain", n_cells, hopping1=h1, hopping2=h2)
+    >>> spin_ham
+    """
+
+    lattice = _generate_lattice(lattice, n_cells, boundary_condition, neighbour_order=2)
+
+    if isinstance(hopping1, (int, float, complex)):
+        hopping1 = [hopping1]
+    hopping1 = math.asarray(hopping1)
+
+    if isinstance(hopping2, (int, float, complex)):
+        hopping2 = [hopping2]
+    hopping2 = math.asarray(hopping2)
+
+    if isinstance(phi, (int, float, complex)):
+        phi = [phi]
+    phi = math.asarray(phi)
+
+    if hopping1.shape not in [(1,), (lattice.n_sites, lattice.n_sites)]:
+        raise ValueError(
+            f"The hopping1 parameter should be a constant or an array of shape ({lattice.n_sites},{lattice.n_sites})"
+        )
+
+    if hopping2.shape not in [(1,), (lattice.n_sites, lattice.n_sites)]:
+        raise ValueError(
+            f"The hopping2 parameter should be a constant or an array of shape ({lattice.n_sites},{lattice.n_sites})"
+        )
+
+    if phi.shape not in [(1,), (lattice.n_sites, lattice.n_sites)]:
+        raise ValueError(
+            f"The phi parameter should be a constant or an array of shape ({lattice.n_sites},{lattice.n_sites})"
+        )
+
+    spin = 2
+    hopping_ham = 0.0 * FermiWord({})
+    for edge in lattice.edges:
+        i, j, order = edge
+        for s in range(spin):
+            s1 = i * spin + s
+            s2 = j * spin + s
+            if order == 0:
+                if hopping1.shape == (1,):
+                    hopping_ham -= hopping1[0] * (
+                        FermiWord({(0, s1): "+", (1, s2): "-"})
+                        + FermiWord({(0, s2): "+", (1, s1): "-"})
+                    )
+                else:
+                    hopping_ham -= hopping1[i][j] * (
+                        FermiWord({(0, s1): "+", (1, s2): "-"})
+                        + FermiWord({(0, s2): "+", (1, s1): "-"})
+                    )
+                print("order", hopping_ham)
+
+            else:
+                if hopping2.shape == (1,):
```

Comment:
Maybe you could do something like:
```py
hopping_term = hopping2[0] if hopping2.shape == (1,) else hopping2[i][j]
phi_term = phi[0] if phi.shape == (1,) else phi[i][j]
```
This should simplify this loop into one compact statement based on order.